library/data-model/src/databaseEngine/engine.ts

```diff
@@ -0,0 +1,1209 @@
+        revision_id: revisionId,
+        record_id: formRecord.record_id,
+        annotations: formRecord.annotations[fieldName],
+        created: formRecord.updated.toISOString(),
+        created_by: formRecord.updated_by,
+      };
+
+      avpsToCreate.push(avp);
+      avpMap[fieldName] = avpId;
+    }
+
+    // Create all AVPs (with attachment handling if configured)
+    await this.bulkCreateAvps(avpsToCreate);
+
+    return avpMap;
+  }
+
+  /**
+   * Create new AVPs for changed fields, reuse AVP IDs for unchanged fields.
+   * This optimizes storage by only creating new AVPs when data actually changes.
+   *
+   * @param params.formRecord - The validated form record with changes
+   * @param params.parentRevision - The parent revision to compare against
+   * @param params.newRevisionId - The new revision ID being created
+   * @returns Map of field names to AVP IDs (new or reused)
+   * @throws Error if equality function not configured
+   * @throws Error if AVP operations fail
+   */
+  private async createOrReuseAvps({
+    formRecord,
+    parentRevision,
+    newRevisionId,
+  }: {
+    formRecord: FormRecord;
+    parentRevision: ExistingRevisionDBDocument;
+    newRevisionId: string;
+  }): Promise<Record<string, string>> {
+    if (!this.getEqualityFn) {
+      throw new Error(
+        'Form operations not configured. Call configure() first.'
+      );
+    }
+
+    const avpMap: Record<string, string> = {};
+    const avpsToCreate: Array<NewAvpDBDocument> = [];
+
+    // Fetch all parent AVPs efficiently in parallel
+    const parentAvpIds = Object.values(parentRevision.avps);
+    const parentAvpPromises = parentAvpIds.map(id => this.core.getAvp(id));
+    const parentAvps = await Promise.all(parentAvpPromises);
+
+    // Build map of field name -> parent AVP
+    const parentAvpsByField = new Map<string, ExistingAvpDBDocument>();
+    for (const [fieldName, avpId] of Object.entries(parentRevision.avps)) {
+      const avp = parentAvps.find(a => a._id === avpId);
+      if (avp) {
+        parentAvpsByField.set(fieldName, avp);
+      }
+    }
+
+    // Check each field for changes
+    for (const [fieldName, fieldValue] of Object.entries(formRecord.data)) {
+      const fieldType =
+        formRecord.field_types[fieldName] ?? UNKNOWN_TYPE_FALLBACK;
+      const fieldAnnotation = formRecord.annotations[fieldName];
+
+      const parentAvp = parentAvpsByField.get(fieldName);
+
+      // Determine if data or annotations changed
+      const equalityFn = this.getEqualityFn(fieldType);
+
+      const hasDataChanged =
+        parentAvp === undefined
+          ? true
+          : !(await equalityFn(parentAvp.data, fieldValue));
+
+      const hasAnnotationChanged =
+        parentAvp === undefined
+          ? true
+          : !(await equalityFn(parentAvp.annotations, fieldAnnotation));
+
+      if (hasDataChanged || hasAnnotationChanged) {
+        // Create new AVP
+        const avpId = generateAvpID();
```

Comment:
Here we're creating the id inside the creator function, which I think is the right thing to do